### ISSUE TITLE
Support static methods with @Redirect

### ISSUE THREAD

**Issue by kashike** [SpongePowered/Mixin#79]
According to @Zidane this should be fine, though it it not currently supported.

Chat:
```
05:38:13        +kashike | I'm having so much fun myself                                                             
05:38:19        +kashike | InvalidInjectionException: 'static' modifier of callback method does not match target     
05:38:22        +kashike | when I add static                                                                         
05:38:23        +kashike | InvalidMixinException Mixin classes cannot contain visible static methods or fields,      
05:38:26        +kashike | yay.                                                                                      
05:38:33         @Zidane | Well no duh kashike                                                                       
05:38:44         @Zidane | Why in the hell would you put a public static in a mixin?                                 
05:38:51         @Zidane | Mixins DO NOT EXIST                                                                       
05:39:11         @Zidane | So how would MixinMyThingy.Thing work :P                                                  
05:39:25         @Zidane | Just make it private static                                                               
05:39:39     kenzierocks | what if you want to mix into a static method, might be the thing?                         
05:39:48        +kashike | yup                                                                                       
05:40:05     kenzierocks | i think you can leave it as a regular method and it'll work                               
05:40:18        +kashike | 05:38:19        +kashike | InvalidInjectionException: 'static' modifier of callback method does not match target                                                                     
05:40:48     kenzierocks | oh,fun                                                                                    
05:41:51         @Zidane | kashike, you'll need to do those as overwrites                                            
5:41:51         @Zidane | kashike, you'll need to do those as overwrites                                    
5:41:54         @Zidane | Which you are allowed                                                             
5:42:01         @Zidane | Look at MixinDimensionManager                                                     
5:42:06         @Zidane | in SpongeForge                                                                    
5:42:09     kenzierocks | why                                                                               
5:42:31         @Zidane | You cannot have public static methods in a mixin                                  
5:42:36         @Zidane | Read what I said above                                                            
5:42:37         @Zidane | UNLESS                                                                            
5:42:40         @Zidane | Its an Overwrite                                                                  
5:43:03        +kashike | well that's annoying, I didn't want to have to replace the entire method          
5:43:24        +kashike | Aaron1011: ^                                                                      
5:43:27     kenzierocks | ...that's the why of why doesn't it work                                          
5:43:33     kenzierocks | what's the why of why it's not allowed                                            
5:43:42         @Zidane | It doesn't work because its a safeguard                                           
5:43:43     kenzierocks | besides that yes, you can't call static methods, duh.                             
5:43:47         @Zidane | and it makes NO sense                                                             
5:43:54     kenzierocks | why not @Inject on a static                                                       
5:43:57         @Zidane | Don't ping Aaron earlier                                                          
5:44:04         @Zidane | Mumfrey wrote mixin, not him                                                      
5:44:08         @Zidane | Go make an issue                                                                  
5:44:24        +kashike | he's the one who told me to use @Redirect, so that's why the ping                 
5:45:22      gravityfox | hey, what's the best way to get the root world directory?                         
5:45:41      gravityfox | cause they are different on server and client and I want to support both properly 
5:45:51        +kashike | Game#getSavesDirectory                                                            
5:46:02      gravityfox | oh okay cool                                                                      
5:46:30        +kashike | Zidane: thanks, saved me the time trying to get this to work as it was            
5:46:42         @Zidane | kashike, make an issue on Mixin                                                   
5:46:46         @Zidane | Redirect ought to be fine                                                         
5:46:55         @Zidane | (Unless Mumfrey thanks otherwise)                                                 
5:46:57        +kashike | Alright                                                                           
```

**Comment by Mumfrey** [comment]
`@Redirect` already supports `static`, as do all injectors. If your target is `static` then your callback must also be `static`, but must not be `public` because Mixins *cannot have visible `public static` members* as you were already explained in the IRC log above.

The `static` modifier must match the staticness of your **target** method (the caller), the staticness of the method call **being redirected** (the callee) is irrelevant, you simply will (or will not) need to prepend a variable matching the type of the callee to your handler's args. For `static` callees you obviously **do not** need to do this, because there is no instance that the method is being invoked upon, and thus no instance to receive.

Simply make your callback `private static` and that's all you need to do. 

To summarise:

* `static` modifier on your callback must match `static` modifier of the **target** (caller)
* callback **must** be `private`
* if redirected method (callee) is `static`, then callback signature matches callee's. If it's an instance method, then the signature must have the instance class prepended to the method arguments
* you cannot have visible `static` members of any kind on a Mixin because there is no conceivable reason to do so, the only exception (as already explained in the IRC log) is `@Overwrite` for obvious reasons, because the destination method already exists and is thus callable.

HTH

Closing because invalid.

**Comment by Mumfrey** [comment]
Just to clarify further, here's a canned example:

```java
public class SomeTargetClass {

    private static String state = "foo";
    private static Logger myLogger = LogManager.getLogger();
    
    public static method aTargetMethod(String text) {
        // This is the call we want to redirect
        SomeTargetClass.logger.info("[{}] {}", SomeTargetClass.state, text);
    }
}
```

And here is an `@Redirect` example to redirect the call to the `info` method:

```java
@Mixin(SomeTargetClass.class)
public abstract class MyMixin {

    @Redirect(method = "aTargetMethod",
        at = @At(
            value = "INVOKE", 
            target = "Lorg/apache/logging/log4j/Logger;info(Ljava/lang/String;[Ljava/lang/Object;)V",
            ordinal = 0
        )
    )
    private static void info(Logger logger, String message, Object... params) {
        //                   ^ notice the callback receives the instance of the
        //                     callee as the first argument

        // All we're actually doing is changing the call to info to a call to warn,
        // but we could do anything we like here, including doing nothing at all
        logger.warn(message, params);
    }
}
